### PR TITLE
add CTC batched beam search

### PR DESCRIPTION
**Pull request by lilithgrigoryan**
# What does this PR do ?

Batched beam search for CTC with NGPU-LM (n-gram LM on GPU).

Results for `nvidia/parakeet-ctc-1.1b` on SLURP (out-of-domain):

| Decoding Method    | Beam Size | Beam Beta | LM Weight* | SLURP Test WER (%) ↓ | RTFx ↑  |
|--------------------|-----------|------------|------------|-----------------------|---------|
| greedy                     | –                | –                 | –                  | 19.93                             | 484       |
| beam_batch           | 8               | 1.0              | 0.6              | 15.84                      | 439       |
| beam_batch           | 12              | 1.0              | 0.6               | 15.75                       | 432       |
| beam_batch           | 16              | 1.0              | 0.6               | 15.72                       | 427       |

\*LM weight - optimal on SLURP dev set with step `0.1`

Batch size 32, high fp32 matmul precision, sorted manifest, A5000 GPU.
6-gram LM is built on SLURP train texts.

**Collection**: [ASR]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

## Usage without LM

```shell
python examples/asr/speech_to_text_eval.py \
   pretrained_name="nvidia/parakeet-ctc-1.1b" \
   dataset_manifest=<dataset_manifest>  \
   batch_size=32 \
   output_filename=decoded.jsonl \
   ctc_decoding.strategy="beam_batch" \
   ctc_decoding.beam.beam_size=12 \
   ctc_decoding.beam.beam_beta=0.3
```

## Usage with LM
**Step 1: Build LM for ASR Model (tokenizer-dependent) on domain texts**

```shell
python nemo/scripts/asr_language_modeling/ngram_lm/train_kenlm.py \
      nemo_model_file="nvidia/parakeet-ctc-1.1b" \
      train_paths=["<train_manifest>"] \
      kenlm_bin_path=$KENLM_BIN_PATH \
      kenlm_model_file=parakeet-ctc-1.1b_lm-o6.arpa \
      ngram_length=6 \
      preserve_arpa=true \
      save_nemo=true
```

**Step 2: Run Decoding with LM**

```shell
python examples/asr/speech_to_text_eval.py \
   pretrained_name="nvidia/parakeet-ctc-1.1b" \
   dataset_manifest=<dataset_manifest>  \
   batch_size=32 \
   output_filename=decoded.jsonl \
   ctc_decoding.strategy="beam_batch" \
   ctc_decoding.beam.ngram_lm_model="parakeet-ctc-1.1b_lm-o6.arpa.nemo" \
   ctc_decoding.beam.ngram_lm_alpha=0.4 \
   ctc_decoding.beam.beam_beta=0.4 \
   ctc_decoding.beam.beam_size=12
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Comment by ko3n1g** [comment]
There’s a known error, let me know if you want to proceed with merging this PR

File: nemo/collections/asr/parts/utils/batched_beam_decoding_utils.py [hunk 1]
```diff
@@ -535,3 +534,346 @@ def flatten_sort_(self, score_norm: bool = True):
         self.transcript_hash.copy_(torch.gather(self.transcript_hash, dim=-1, index=indices))
         if self.store_prefix_hashes:
             self.transcript_prefix_hash.copy_(torch.gather(self.transcript_prefix_hash, dim=-1, index=indices))
+
+
+class BatchedBeamHypsCTC:
```
Comment (artbataev) [inline nemo/collections/asr/parts/utils/batched_beam_decoding_utils.py:539]: What is the purpose for separating this class and `BatchedBeamHyps`?

File: nemo/collections/asr/parts/submodules/ctc_batched_beam_decoding.py [hunk 2]
```diff
@@ -0,0 +1,700 @@
...
+class BatchedBeamCTCComputer(WithOptionalCudaGraphs, ConfidenceMethodMixin):
+    """
+    Batched Alignment-Length Synchronous Decoding implementation. Callable.
+    Based on https://ieeexplore.ieee.org/document/9053040 with the following modficiations:
+        - does not support prediction network caching
+        - does not employ transcript length estimation, instead, limits the number of expansions for every frame.
+    """
+
+    INITIAL_MAX_TIME = 375  # initial max time, used to init state for Cuda graphs
+    CUDA_PROGRAM_NAME = b"while_malsd_batch_conditional_rnnt.cu"
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/ctc_batched_beam_decoding.py:166]:
```suggestion
    CUDA_PROGRAM_NAME = b"while_beam_batch_conditional_ctc.cu"
```

File: nemo/collections/asr/parts/submodules/ctc_batched_beam_decoding.py [hunk 3]
```diff
@@ -0,0 +1,700 @@
...
+        # step 2.5: masking inactive hypotheses, updating + recombining batched beam hypoteses
+        torch.where(self.state.active_mask, next_labels, self.state.MINUS_ONE_TENSOR, out=next_labels)
+        self.state.batched_hyps.add_results_no_checks_(next_indices, next_labels, next_scores)
+        self.state.batched_hyps.recombine_hyps_()
+
+        # step 2.6: updating frame idx and active masks
+        self.state.curr_frame_idx.add_(1)
+        torch.greater_equal(self.state.last_timesteps, self.state.curr_frame_idx, out=self.state.active_mask)
+        torch.any(self.state.active_mask, out=self.state.active_mask_any)
+        torch.cuda.set_sync_debug_mode(0)
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/ctc_batched_beam_decoding.py:679]: Please, remove this statement

File: nemo/collections/asr/parts/submodules/ctc_batched_beam_decoding.py [hunk 4]
```diff
@@ -0,0 +1,700 @@
...
+            eos_score = self.ngram_lm_batch.get_final(self.state.batch_lm_states).view(
+                self.state.batched_hyps.scores.shape
+            )
+            self.state.batched_hyps.scores += eos_score * self.beam_alpha
+
+    def __call__(
+        self,
+        x: torch.Tensor,
+        out_len: torch.Tensor,
+    ) -> Tuple[rnnt_utils.BatchedHyps, Optional[rnnt_utils.BatchedAlignments], Any]:
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/ctc_batched_beam_decoding.py:696]: return type - `BatchedBeamHypsCTC` or `BatchedBeamHyps`

**Review by artbataev — CHANGES_REQUESTED** [review]
Overall looks good! Minor comments.
Suggestion: merge `BatchedBeamHypsCTC` and `BatchedBeamHyps` if possible

File: nemo/collections/asr/parts/submodules/ctc_batched_beam_decoding.py [hunk 5]
```diff
@@ -0,0 +1,700 @@
...
+        beam_alpha: float = 1.0,
+        beam_beta: float = 0.0,
+        beam_threshold: float = 20.0,
+        kenlm_path: str = None,
+        allow_cuda_graphs: bool = True,
+    ):
+        """
+        Init method.
+        Args:
+            decoder: Prediction network from RNN-T
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/ctc_batched_beam_decoding.py:195]: 🤔

File: nemo/collections/asr/parts/submodules/ctc_batched_beam_decoding.py [hunk 6]
```diff
@@ -0,0 +1,700 @@
...
+    ngram_lm_batch: Optional[NGramGPULanguageModel]
+
+    def __init__(
+        self,
+        blank_index: int,
+        beam_size: int,
+        return_best_hypothesis: bool = True,
+        preserve_alignments=False,
+        compute_timestamps: bool = False,
+        beam_alpha: float = 1.0,
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/ctc_batched_beam_decoding.py:186]: can we unify this parameter with all other models and use `ngram_lm_alpha`?

File: nemo/collections/asr/parts/submodules/ctc_batched_beam_decoding.py [hunk 7]
```diff
@@ -0,0 +1,700 @@
...
+        self.beam_threshold = beam_threshold
+
+        assert not self.preserve_alignments, "Preserve aligments is not supported"
+        assert not self.compute_timestamps, "Compute timestamps is not supported"
+
+        self.state = None
+        self.full_graph = None
+        self.separate_graphs = None
+
+        # prprprprpr
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/ctc_batched_beam_decoding.py:230]: what?

File: nemo/collections/asr/parts/submodules/ctc_batched_beam_decoding.py (hunk 3 above)
Comment (lilithgrigoryan) [reply nemo/collections/asr/parts/submodules/ctc_batched_beam_decoding.py]: Done. Thanks

File: nemo/collections/asr/parts/utils/batched_beam_decoding_utils.py (hunk 1 above)
Comment (lilithgrigoryan) [reply nemo/collections/asr/parts/utils/batched_beam_decoding_utils.py]: We have a single class now. Thanks!

File: nemo/collections/asr/parts/submodules/ctc_batched_beam_decoding.py (hunk 6 above)
Comment (lilithgrigoryan) [reply nemo/collections/asr/parts/submodules/ctc_batched_beam_decoding.py]: Done. Thank you

File: nemo/collections/asr/parts/submodules/ctc_batched_beam_decoding.py [hunk 8]
```diff
@@ -0,0 +1,700 @@
...
+        self.compute_timestamps = compute_timestamps
+        self.allow_cuda_graphs = allow_cuda_graphs
+        self.return_best_hypothesis = return_best_hypothesis
+
+        self.beam_alpha = beam_alpha
+        self.beam_beta = beam_beta
+        self.beam_threshold = beam_threshold
+
+        assert not self.preserve_alignments, "Preserve aligments is not supported"
+        assert not self.compute_timestamps, "Compute timestamps is not supported"
```
[the remark this answers is not in the document]
Comment (lilithgrigoryan) [reply nemo/collections/asr/parts/submodules/ctc_batched_beam_decoding.py]: Added support for timestamps

File: nemo/collections/asr/parts/submodules/ctc_batched_beam_decoding.py (hunk 7 above)
Comment (lilithgrigoryan) [reply nemo/collections/asr/parts/submodules/ctc_batched_beam_decoding.py]: removed:)

File: audio
Comment (artbataev) [inline audio:1]: Please remove

File: nemo/collections/asr/parts/submodules/ctc_beam_decoding.py [hunk 9]
```diff
@@ -877,6 +878,116 @@ def _k2_decoding(self, x: torch.Tensor, out_len: torch.Tensor) -> List['WfstNbes
...
+
+        self.ngram_lm_alpha = ngram_lm_alpha
+        self.beam_beta = beam_beta
+        self.beam_threshold = beam_threshold
+
+        # Default beam search args
+        self.ngram_lm_model = ngram_lm_model
+
+        # Default beam search scorer functions
+        self.default_beam_scorer = None
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/ctc_beam_decoding.py]: Are these variables necessary?

**Review by artbataev — APPROVED** [review]
Thanks a lot! Approved. 🚀

File: nemo/collections/asr/parts/submodules/ctc_beam_decoding.py (hunk 9 above)
Comment (lilithgrigoryan) [reply nemo/collections/asr/parts/submodules/ctc_beam_decoding.py]: Removed, thank you
